roles/openshift_health_checker/openshift_checks/sdn.py

```diff
@@ -0,0 +1,698 @@
+                self.register_file('resolv.conf', None, '/etc/resolv.conf')
+                self.save_command_output('modules', ['lsmod'])
+                self.save_command_output('sysctl', ['sysctl', '-a'])
+                # XXX: Would self.get_var('openshift', 'docker') suffice?
+                self.save_command_output('docker-version',
+                                         ['docker', 'version'])
+                # XXX: Would saving the output of the oc_version module
+                # suffice?
+                self.save_command_output('oc-version', ['oc', 'version'])
+                self.register_file('os-version', None,
+                                   '/etc/system-release-cpe')
+            except OpenShiftCheckException as exc:
+                self.register_failure(exc)
+
+        group_names = self.get_var('group_names', default=[])
+        if 'oo_masters_to_config' in group_names:
+            self.check_master()
+        if 'oo_nodes_to_config' in group_names:
+            self.check_node()
+        return {}
+
+    def save_command_output(self, path, command):
+        """Execute the provided command using the command module
+        and save its output to the specified file.
+
+        If the command is a string, use a shell.  Otherwise, assume the command
+        is a list, join it with spaces, and execute it without shell.
+        """
+        self.register_file(path, self.read_command_output(command))
+
+    def read_command_output(self, command):
+        """Execute the provided command using the command module
+        and return its output.
+
+        If the command is a string, use a shell.  Otherwise, assume the command
+        is a list, join it with spaces, and execute it without shell.
+        """
+        uses_shell = False
+        if isinstance(command, six.string_types):
+            uses_shell = True
+        else:
+            command = ' '.join(command)
+
+        command_args = dict(_raw_params=command, _uses_shell=uses_shell)
+        # Use self._execute_module instead of self.execute_module because
+        # the latter sets self.changed.
+        result = self._execute_module('command', command_args)
+        if result.get('rc', 0) != 0 or result.get('failed'):
+            raise OpenShiftCheckException(
+                'RemoteCommandFailure',
+                'Failed to execute command on remote host: %s' % command)
+
+        return result['stdout']
+
+    def check_master(self):
+        """Gather diagnostic information on a master and ensure it can connect
+        to kubelets."""
+        if self.want_full_results:
+            conf_base_path = self.get_var('openshift.common.config_base')
+            master_conf_path = os.path.join(conf_base_path, 'master',
+                                            'master-config.yaml')
+            self.register_file('master-config.yaml', None, master_conf_path)
+
+            # XXX: Loading master logs can take in inordinate amount of
```

Comment:
Informal testing on free-int shows that `journalctl --boot -n 100` takes about a minute (55 to 58 seconds).  Meanwhile, `journalctl --boot | tail -n 100` just hangs.  Controlling it with a variable makes some sense, but maybe it would be better left to other tools.

Reply:
free-int is a pathological test case. You should expect it to be crazy slow.

`--boot -n 100` isn't the right way to limit the logs anyway. The point of using `--boot` was that we wanted to see the logs from startup, because important things happen at startup. If we have to get only a subset of logs, then we'd want to do something like "the first 5 minutes of logs after the service started and then the most recent 5 minutes of logs". Definitely want to use time rather than number of lines, since we don't know what debug level they're using.

Reply:
Does that mean that for each service, we should get its ExecMainStartTimestamp and then run `journalctl --boot --since=<starttime> --until=<starttime+5m>`, and then do a `journalctl --boot --since=-5m`?

Note that getting the data using Ansible means base64-encoding the output, transferring it over the network, and storing it in a dict, so might still need some limiting `-n` in order to avoid grabbing too much data (say > 1 MiB).

Reply:
You don't need `--boot` if you're explicitly specifying times, but yeah, that looks right. (Assuming the service has been running for more than 5 minutes...)

> Note that getting the data using Ansible means base64-encoding the output, transferring it over the network, and storing it in a dict, so might still need some limiting -n in order to avoid grabbing too much data (say > 1 MiB).

If you limit the data then it's not going to solve the problem it's supposed to be solving, so we'd need to figure out some other way.

Reply:
Changed to save 5 minutes of journal entries from the service's start time, the last 5 minutes of the service's logs, and the last 5 minutes of journal entries.